### PR TITLE
feat: introduce custom reference date to date picker (14.10)

### PR DESCRIPTION
**Pull request by czp13**
## Description

The customer agreed to sponsor the reference date feature and backport it from our latest flow-components version:

Related support ticket: 
- https://support.vaadin.com/projects/VS/queues/custom/3/VS-4437

Related github ticket:
- https://github.com/vaadin-component-factory/enhanced-date-picker/issues/16

The issues in the customer's words:

> If I set the parser pattern "ddMMyy" and the formatting pattern "dd.MM.yyyy" to the enhanced-date-picker, the following input "101070" is parsed to "10.10.2070". I think it makes more sense to parse this input to "10.10.1970" in context of date of birth, for example.
> Is there a way to manually set the breakpoint where the parser switches from 2000 to 1900?
> 

Fixes # (issue)
- Added prepare-frontend maven goal to datepicker so now the mvn build will be successful and the application can be run on the local machine, otherwise, the frontend was not ready and some assets were not bundled to the Jetty web server,
- added NPMPackage "date-fns" , this is needed as we build open this package the feature,
- added reference date feature to DatePicker Java class,
- added reference date feature to datePicker connector JS class, also parsing changed to use date-fns parsing methods,

## Type of change

- [ ] Bugfix
- [x] Feature

## Checklist

- [x] I have read the contribution guide: https://vaadin.com/docs/latest/contributing/overview
- [x] I have added a description following the guideline.
- [x] The issue is created in the corresponding repository and I have referenced it.
- [x] I have added tests to ensure my change is effective and works as intended.
- [x] New and existing tests are passing locally with my change.
- [x] I have performed self-review and corrected misspellings.

#### Additional for `Feature` type of change

- [x] Enhancement / new feature was discussed in a corresponding GitHub issue and Acceptance Criteria were created.

### CODE REVIEW

**Review by web-padawan — CHANGES_REQUESTED** [review]
Thanks for the PR. This change was a new feature and IMO it is not feasible to land in a patch release.
Also, implementing it would require a corresponding change in the web component first.

We need to agree if this is something that we want to be part of Vaadin 14.10

**Comment by TatuLund** [comment]
> We need to agree if this is something that we want to be part of Vaadin 14.10

@web-padawan Yes, you are correct. The branch is wrong, I just noticed it myself too. The simple reason is that 14.10 branch has not been created yet.

**Comment by ZheSun88** [comment]
I removed the `target/14.9`  label. 
after this PR #4441  you can push to 14.10 branch.

**Comment by web-padawan** [comment]
This PR still needs a change in the web component to be implemented first (needs a minor bump).

**Comment by czp13** [comment]
Executed another build as the last build was lost from TeamCity, and picked up this one today again.

**Comment by czp13** [comment]
Rebase and some other related changes happened, let's see the result of this build.

**Comment by czp13** [comment]
Locally the tests are good, so quiet weird behavior, I think something about the run config, or locale is not matching with my dev environment vs the build server:
image

**Comment by ZheSun88** [comment]
Hi @czp13 , the failed test in the validation is running under bower mode, did you run that in your local? 

i just tried on my side, seems getting the same failure.

**Comment by czp13** [comment]
> Hi @czp13 , the failed test in the validation is running under bower mode, did you run that in your local?
> ...
Thank you vm @ZheSun88! I hit the wall so hard with this one, but this can push me forward I feel/hope.
I used this command: 
` mvn -pl vaadin-date-picker-flow-integration-tests clean jetty:run`
And was scratching my hand about the difference between team city build agent and my env, but I will try with bower too, just remember I had some issues with it... I will get back to this.

**Comment by czp13** [comment]
I can confirm with your bower environment variable I can reproduce the bugs in my local environment, finally, I can move forward with this one. Thank you 🙇 🙏 (Just to not be easy, we need to use JDK 18 and also pom.xml needed to be changed, to remove the deprecated JVM param, but apart from that it works)

**Comment by czp13** [comment]
I figured out the bower part with the help of @ZheSun88.
Fingers crossed this will be good now, then I will clean up a bit, and THEN hopefully good to go (merge).

**Review by sissbruecker — CHANGES_REQUESTED** [review]

File: vaadin-date-picker-flow-parent/vaadin-date-picker-flow-demo/pom.xml [hunk 1]
```diff
@@ -1,5 +1,6 @@
 <?xml version="1.0" encoding="UTF-8" standalone="yes"?>
-<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
```
[the remark this answers is not in the document]
Comment (czp13) [reply vaadin-date-picker-flow-parent/vaadin-date-picker-flow-demo/pom.xml]: Done, thanks.

File: vaadin-date-picker-flow-parent/vaadin-date-picker-flow-integration-tests/src/main/java/com/vaadin/flow/component/datepicker/DatePickerViewDemoPage.java [hunk 2]
```diff
@@ -245,6 +245,9 @@ private void createLocaleChangeDatePicker() {
 
         datePicker.addValueChangeListener(
                 event -> updateMessage(message, datePicker));
+        DatePicker.DatePickerI18n i18n = new DatePicker.DatePickerI18n();
+        i18n.setReferenceDate(LocalDate.of(1980, 2, 2));
+        datePicker.setI18n(i18n);
```
[the remark this answers is not in the document]
Comment (czp13) [reply vaadin-date-picker-flow-parent/vaadin-date-picker-flow-integration-tests/src/main/java/com/vaadin/flow/component/datepicker/DatePickerViewDemoPage.java:250]: Thanks for this as well.

Added all tests I could, which was not trivial not just back-porting as some pages/IT classes were looking differently, and other things were missing too, or needed to add, changes, but I believe now we covered the cases which we had in the original PR.

File: vaadin-date-picker-flow-parent/vaadin-date-picker-flow/src/main/java/com/vaadin/flow/component/datepicker/DatePicker.java [hunk 3]
```diff
@@ -53,9 +55,14 @@
...
 @JsModule("./date-picker-datefns.js")
-@JsModule("./datepickerConnector.js")
 @JavaScript("frontend://date-picker-datefns.js")
+@NpmPackage(value = "date-fns", version = "2.29.3")
+@JsModule("./datepickerConnector.js")
 @JavaScript("frontend://datepickerConnector.js")
+@NpmPackage(value = "@vaadin/vaadin-date-picker", version = "4.5.0")
+@JsModule("@vaadin/vaadin-date-picker/src/vaadin-date-picker.js")
+@JsModule("date-fns/")
+// @HtmlImport("frontend://bower_components/vaadin-date-picker/src/vaadin-accordion.html")
```
[the remark this answers is not in the document]
Comment (czp13) [reply vaadin-date-picker-flow-parent/vaadin-date-picker-flow/src/main/java/com/vaadin/flow/component/datepicker/DatePicker.java]: Good point, I wanted to backport this bit as well, as it was part of the backported PR, but it was not working for bower, and then I spent quite many time to figure out what is going on, I saw this workaround with `vaadin-date-picker-flow-parent/vaadin-date-picker-flow/date-fns` and the only reason I could fix the PR that I understood what we are doing here (needed to change JS classes to use this directory date-fns version not from NPM, etc...

File: vaadin-date-picker-flow-parent/vaadin-date-picker-flow/src/main/java/com/vaadin/flow/component/datepicker/GeneratedVaadinDatePicker.java [hunk 4]
```diff
@@ -240,7 +240,7 @@
  * </p>
  * <p>
  * In addition to {@code <vaadin-date-picker>} itself, the following internal
- * components are themable:
+ * components are themeable:
```
[the remark this answers is not in the document]
Comment (czp13) [reply vaadin-date-picker-flow-parent/vaadin-date-picker-flow/src/main/java/com/vaadin/flow/component/datepicker/GeneratedVaadinDatePicker.java]: I try to follow the boy-scout rule, meaning try to fix everything I see on the way (like README.md which was fully outdated), but I reverted back this bit for now.

File: vaadin-date-picker-flow-parent/vaadin-date-picker-flow/src/main/resources/META-INF/resources/frontend/datepickerConnector.js [hunk 5]
```diff
@@ -1,189 +1,203 @@
 (function () {
     const tryCatchWrapper = function (callback) {
-        return window.Vaadin.Flow.tryCatchWrapper(callback, 'Vaadin Date Picker', 'vaadin-date-picker-flow');
+        return window.Vaadin.Flow.tryCatchWrapper(callback, 'Vaadin Date Picker');
```
[the remark this answers is not in the document]
Comment (czp13) [reply vaadin-date-picker-flow-parent/vaadin-date-picker-flow/src/main/resources/META-INF/resources/frontend/datepickerConnector.js]: Same here, it was in the PR looking like that, but reverted back! Thanks for noting.

File: vaadin-date-picker-flow-parent/vaadin-date-picker-flow/src/main/resources/META-INF/resources/frontend/datepickerConnector.js [hunk 6]
```diff
@@ -1,189 +1,203 @@
...
+                        }
+                    })
+                );
+
+                /**
+                 * Extracts the basic component parts of a date (day, month and year)
+                 * to the expected format.
+                 * @param {!Date} date
+                 * @return {{day: number, month: number, year: number}}
+                 */
```
[the remark this answers is not in the document]
Comment (czp13) [reply vaadin-date-picker-flow-parent/vaadin-date-picker-flow/src/main/resources/META-INF/resources/frontend/datepickerConnector.js]: Nice catch, thank you, fixed it! (side effect of tons of refactor, trials :) )

File: vaadin-date-picker-flow-parent/vaadin-date-picker-flow/src/main/resources/META-INF/resources/frontend/datepickerConnector.js [hunk 7]
```diff
@@ -1,189 +1,203 @@
...
+                        .replace(/([a-zA-Z]+)/g, "'$1'")
+                        // insert date placeholder
+                        .replace('06', 'dd')
+                        .replace('6', 'd')
+                        // insert month placeholder
+                        .replace('05', 'MM')
+                        .replace('5', 'M')
+                        // insert year placeholder
+                        .replace('1234', 'yyyy');
+                    const isValidPattern = pattern.includes('d') && pattern.includes('M') && pattern.includes('y');
```
[the remark this answers is not in the document]
Comment (czp13) [reply vaadin-date-picker-flow-parent/vaadin-date-picker-flow/src/main/resources/META-INF/resources/frontend/datepickerConnector.js]: Nice one, I fixed this everywhere I see in that class! 🙇

File: vaadin-date-picker-flow-parent/vaadin-date-picker-flow/src/main/resources/META-INF/resources/frontend/datepickerConnector.js [hunk 8]
```diff
@@ -1,189 +1,203 @@
...
-                            return {day: date.getDate(), month: date.getMonth(), year: date.getFullYear()};
-                        }
+                /**
+                 * Parse date string of one of the following date formats:
+                 * - ISO 8601 `"YYYY-MM-DD"`
+                 * - 6-digit extended ISO 8601 `"+YYYYYY-MM-DD"`, `"-YYYYYY-MM-DD"`
+                 * @param {!string} str Date string to parse
+                 * @return {Date} Parsed date
+                 */
+                function _parseDate(str) {
```
[the remark this answers is not in the document]
Comment (czp13) [reply vaadin-date-picker-flow-parent/vaadin-date-picker-flow/src/main/resources/META-INF/resources/frontend/datepickerConnector.js]: Changed this as you wished as well!

[the remark this answers is not in the document]
Comment (czp13) [reply vaadin-date-picker-flow-parent/vaadin-date-picker-flow/src/main/resources/META-INF/resources/frontend/datepickerConnector.js]: And thanks for all the comments, review 👏 🙇

**Review by sissbruecker — APPROVED** [review]

**Pull request merged**